이메일 & 닉네임 중복 체크 API 개발

zlzzlzz2l opened the pull request:
# 이메일 & 닉네임 중복 체크 API 개발

이메일과 닉네임 중복 체크 API를 추가했습니다.

## 이메일 중복 체크
이메일 중복일 때
![image](https://user-images.githubusercontent.com/48669011/146683584-be8c91d5-204c-4691-9341-dbd273d78b33.png)

이메일 중복이 아닐 때
![image](https://user-images.githubusercontent.com/48669011/146683597-9681051a-b848-451f-98c3-1e3ade1f421d.png)

## 닉네임 중복 체크
닉네임 중복일 때
![image](https://user-images.githubusercontent.com/48669011/146683729-a80247a5-4d34-4b0a-945e-a0e7813dcbb0.png)

닉네임 중복이 아닐 때
![image](https://user-images.githubusercontent.com/48669011/146683738-b8365e66-a98f-49fd-9f48-e9df53c61a8f.png)

### 이메일 중복 체크 API
![image](https://user-images.githubusercontent.com/48669011/146683789-ab4f8bbf-0594-4c83-9237-7c17d842134f.png)
![image](https://user-images.githubusercontent.com/48669011/146683797-21141ca7-a3df-4485-a017-0a136abed23f.png)

### 닉네임 중복 체크 API
![image](https://user-images.githubusercontent.com/48669011/146683810-704a2efa-c648-4b67-93ba-3ba056f66e0d.png)
![image](https://user-images.githubusercontent.com/48669011/146683814-e43b23f7-3545-42de-8e68-853e8786f06c.png)

zlzzlzz2l committed:
Update: #10 - 회원 컨트롤러 이메일 중복 체크 메서드 추가

zlzzlzz2l committed:
Update: #10 - 회원 컨트롤러 이메일 중복 체크 메서드 추가

zlzzlzz2l committed:
Update: #10 - 회원 컨트롤러 닉네임 중복 체크 메서드 추가

zlzzlzz2l committed:
Update: #10 - 회원 컨트롤러 닉네임 중복 체크 메서드 추가

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 회원가입 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 회원가입 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 구현

zlzzlzz2l committed:
Create: #10 - 회원 서비스 테스트 메서드 생성

zlzzlzz2l committed:
Create: #10 - 회원 서비스 테스트 메서드 생성

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 회원가입 메서드명 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 회원가입 메서드명 수정

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 회원가입 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 회원가입 메서드 구현

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드명 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드명 수정

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드 구현

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드명 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드명 수정

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드 구현

zlzzlzz2l committed:
Update: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드 구현

zlzzlzz2l committed:
Refactor: #10 - 회원 컨트롤러 닉네임 중복 체크 메서드 버그 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 컨트롤러 닉네임 중복 체크 메서드 버그 수정

0xe82de commented on src/main/java/com/fourfourfour/damoa/controller/user/UserController.java:
```diff
@@ -149,4 +149,32 @@ public ResponseEntity register(final @Valid @RequestBody UserDto user, Errors er
+            @ApiResponse(responseCode = "404", description = "이메일이 중복되지 않습니다"),
+            @ApiResponse(responseCode = "500", description = "서버 장애 발생")
+    })
+    @PostMapping("/email/{userEmail}/exists")
```
@GetMapping으로 바꾸면 어떨까요?

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/main/java/com/fourfourfour/damoa/controller/user/UserController.java:
```diff
@@ -177,4 +177,31 @@ public ResponseEntity checkEmailDuplicate(@PathVariable String userEmail) {
+            @ApiResponse(responseCode = "404", description = "닉네임이 중복되지 않습니다"),
+            @ApiResponse(responseCode = "500", description = "서버 장애 발생")
+    })
+    @PostMapping("/nickname/{userNickname}/exists")
```
이 부분도 @GetMapping으로 바꾸면 어떨까요?

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -1,16 +1,40 @@
     @Autowired
     private UserDao userDao;
 
+    @Test
+    public void testInsertUser(){
+        //given
+        UserDto user = new UserDto();
+        user.setUserEmail("test@test.com");
+        user.setUserPw("1234");
+        user.setUserJob("대학생");
+        user.setUserGender("female");
+        user.setUserNickname("테스트");
+        user.setUserPromotionAgree(true);
+        user.setUserPrivacyAgree(true);
+        user.setUserLocationAgree(true);
+        user.setUserServiceAgree(true);
+        //when
+
+        userDao.insertUser(user);
+
+        //then
+        String findNickname = userDao.selectByNickname(user.getUserNickname());
+        Assertions.assertThat(user.getUserNickname()).isEqualTo(findNickname);
+    }
+
```
회원 가입 전 회원 객체와 Dao로 가져온 회원 객체를 비교하는 게 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -37,4 +37,16 @@ public void testInsertUser(){
         Assertions.assertThat(user.getUserNickname()).isEqualTo(findNickname);
     }
 
+    @Test
+    public void testSelectEmailByEmail(){
+        //given
+        String userEmail = "test4@test.com";
+
+        //when
+        String result = userDao.selectEmailByEmail(userEmail);
+
+        //then
+        Assertions.assertThat(result).isNull(); // Null일 때 통과
+    }
+
```
위 로직 밑에서 DB에 회원 정보를 넣은 다음 중복 체크하는 로직을 추가하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -49,4 +49,15 @@ public void testSelectEmailByEmail(){
         Assertions.assertThat(result).isNull(); // Null일 때 통과
     }
 
+    @Test
+    public void testSelectByNickname(){
+        //given
+        String userNickname = "test4";
+
+        //when
+        String result = userDao.selectByNickname(userNickname);
+
+        //then
+        Assertions.assertThat(result).isNull(); // Null일 때 통과
+    }
```
이메일 중복 테스트 코드와 마찬가지로 위 로직 밑에서 회원 가입 후 닉네임 중복 체크 로직을 추가하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/service/user/UserServiceImplTest.java:
```diff
@@ -1,18 +1,42 @@
     @Test
-    void register() {
     public void testRegister() {
+        //given
+        UserDto user = new UserDto();
+        user.setUserEmail("test@test.com");
+        user.setUserPw("1234");
+        user.setUserJob("대학생");
+        user.setUserGender("female");
+        user.setUserNickname("테스트");
+        user.setUserPromotionAgree(true);
+        user.setUserPrivacyAgree(true);
+        user.setUserLocationAgree(true);
+        user.setUserServiceAgree(true);
+
+        //when
+        userService.register(user);
+
+        //then
+        // 서비스 단위에서 이 부분이 필요한지는 좀 더 공부해야 될 느낌입니다..!!
+        boolean resultEmail = userService.isEmailDuplication(user.getUserEmail());
+        Assertions.assertThat(resultEmail).isFalse(); // False라면 중복되지 않기 때문에 가입 가능
+        boolean resultNickname = userService.isEmailDuplication(user.getUserNickname());
+        Assertions.assertThat(resultNickname).isFalse(); // False라면 중복되지 않기 때문에 가입 가능
```
Dao 테스트와 마찬가지로 회원 가입 전의 객체와 Dao에서 가져온 객체를 비교하는 로직을 추가해야 할 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/service/user/UserServiceImplTest.java:
```diff
@@ -41,6 +41,11 @@ public void testRegister() {
 
     @Test
     public void testIsEmailDuplication() {
+        String userEmail = "test3@test.com";
+
+        boolean result = userService.isEmailDuplication(userEmail);
+
+        Assertions.assertThat(result).isTrue(); // True일 때 중복
```
Dao 테스트 코드 코멘트와 마찬가지로 회원 정보를 DB에 넣은 후 중복 체크하는 로직을 추가하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/service/user/UserServiceImplTest.java:
```diff
@@ -50,5 +50,10 @@ public void testIsEmailDuplication() {
 
     @Test
     public void testIsNicknameDuplication() {
+        String userNickname = "testtest11";
+
+        boolean result = userService.isNicknameDuplication(userNickname);
+
+        Assertions.assertThat(result).isTrue(); // True일 때 중복
```
Dao 테스트 코드 코멘트와 마찬가지로 회원 정보를 DB에 넣은 후 중복 체크하는 로직을 추가하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

Dao와 서비스 테스트에서 이메일과 닉네임 중복 체크 메서드는 이메일이나 닉네임 중복 체크 후, 회원가입을 진행하는 로직으로 구성했습니다.

0xe82de commented:
api 매핑 경로를 "api/v1/users" 로 수정해주세요~!

zlzzlzz2l committed:
Refactor: #10 - 회원 컨트롤러 API 매핑 경로 수정

zlzzlzz2l commented:
> api 매핑 경로를 "api/v1/users" 로 수정해주세요~!

yjs1213 reviewed (approved):
고생하셨습니다~

zlzzlzz2l committed:
Merge branch 'develop' into feature/user/#10

0xe82de reviewed (changes requested):
코멘트입니다.

zlzzlzz2l committed:
Merge branch 'feature/user/#10' of https://github.com/four-four-four/DAMOA-Backend into feature/user/#10

zlzzlzz2l committed:
Update: #10 - 회원 Dao 회원 탐색 메서드 선언부 추가

zlzzlzz2l committed:
Update: #10 - 회원 Dao 회원 탐색 메서드 구현

테스트할 때 쓰일 회원 탐색 메서드를 구현했습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 회원가입 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 회원가입 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드 수정

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -5,36 +5,42 @@
-        Assertions.assertThat(user.getUserNickname()).isEqualTo(findNickname);
+       //then
+        UserDto result = userDao.selectUser(user.getUserIdx());
+        Assertions.assertThat(user).isEqualTo(result);
```
다음 테스트 메서드 작성할 때를 위해서 `Assertions`를 `static import`로 빼면 좋을 것 같아요.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -46,13 +46,29 @@ public void testInsertUser(){
 
         //when
-        String result = userDao.selectEmailByEmail(userEmail);
+        String result = userDao.selectEmailByEmail(user.getUserEmail()); // 이메일 중복 체크
```
`result`가 `null` 인지 체크하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -46,13 +46,29 @@ public void testInsertUser(){
+        }
+        else {
+            Assertions.assertThat(result).isNotNull(); // Not null일 때 중복
+        }
```
윗 부분 로직에서 `result`가 `null`일 때 회원 가입을 진행하고, 그 뒤에 다시 이메일 중복체크를 위해 `userDao.selectEmail()` 메서드를 호출했을 때 `null`이 아닌지 체크하는게 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -74,12 +74,30 @@ public void testSelectEmailByEmail(){
 
         //when
-        String result = userDao.selectByNickname(userNickname);
+        String result = userDao.selectByNickname(user.getUserNickname()); // 이메일 중복 체크
```
주석이 잘못 작성된 것 같네요.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/dao/user/JdbcUserDaoImplTest.java:
```diff
@@ -74,12 +74,30 @@ public void testSelectEmailByEmail(){
+        }
+        else {
+            Assertions.assertThat(user).isNotNull(); //Not null일 때 중복
+        }
```
이 테스트 메서드도 이메일 테스트 메서드 코멘트와 같이 수정하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/service/user/UserServiceImplTest.java:
```diff
@@ -48,11 +48,32 @@ public void testRegister() {
+        }
+        else {
+            Assertions.assertThat(result).as("이메일이 중복됩니다.").isFalse(); // false가 아닐 때 중복
+        }
```
`Dao` 테스트 코멘트와 마찬가지로 `result`가 `false`인지 체크하고, 회원가입을 진행한 뒤에 이메일 중복 체크를 하기 위해 `userService.isEmailDuplication()` 메서드 결과가 `true`인지 체크하도록 로직을 수정하면 좋을 것 같습니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de commented on src/test/java/com/fourfourfour/damoa/model/service/user/UserServiceImplTest.java:
```diff
@@ -78,10 +78,29 @@ public void testIsEmailDuplication() {
+        }
+        else {
+            Assertions.assertThat(result).as("닉네임이 중복됩니다.").isFalse(); // false가 아닐 때 중복
+        }
```
이메일 코멘트와 동일합니다.

zlzzlzz2l replied:
반영했습니다.

0xe82de reviewed (changes requested):
리뷰 확인해주시고 바꿀 게 있으면 바꾸시면 될 것 같아요 :)

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 Assertions static import로 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 Assertions static import로 수정

0xe82de commented on src/main/java/com/fourfourfour/damoa/model/dao/user/UserDao.java:
```diff
@@ -12,5 +12,6 @@
     String selectEmailByEmail(String userEmail);
     String selectByNickname(String userNickName);
     UserDto selectByUserEmail(String userEmail);
+    UserDto selectUser(long integer);
```
테스트 코드에서 `user` 객체를 가져오기 위해 `selectUser` 메서드를 추가하신 것 같은데 바로 위에 `selectByUserEmail` 메서드로 `user` 객체를 가져올 수 있습니다. `userIdx`로 `user` 객체를 가져와야 하는 이유가 있는게 아니라면 기존의 메서드를 활용하는게 좋을 것 같습니다~ 만약 `selectByUserEmail` 메서드를 사용하신다면 메서드 이름을 `selectUserByUserEmail`로 리팩토링해주시면 감사합니다~

만약 새롭게 추가하신 `selectUser` 메서드 사용하셔야 한다면 아래 내용을 고려해주세요~!

메서드 이름 : `selectUserByUserIdx`
매개변수 타입 : `Long`
매개변수 이름 : `userIdx`

zlzzlzz2l replied:
selectByEmail 메서드를 이용하는 게 나을 것 같아서 리뷰 토대로 반영했습니다.

0xe82de reviewed (changes requested):
리뷰 추가합니다 :)

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 불필요한 메서드 삭제

selectByUserEmail 메서드를 사용하여 회원 객체를 가져올 수 있기 때문에 selectUser 메서드를 삭제하였습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 회원가입 메서드 수정

selectByUserEmail 메서드로 회원 객체를 가져올 수 있어 수정하였습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 회원가입 메서드 수정

selectByEmail 메서드로 회원 객체를 가져올 수 있어 메서드를 수정하였습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 이메일 & 닉네임 중복 체크 메서드 리턴 타입 변경

이메일 & 닉네임 중복 체크 메서드를 수정하기 위해 변경했습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 이메일 & 닉네임 리턴값 변경

Dao에서 0 or 1을 반환하기 때문에 변경하였습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 이메일 & 닉네임 메서드 리턴 타입 수정

null을 반환할 수 있어 리턴 타입을 Integer로 설정했지만 0 or 1만을 반환하기 때문에 리턴 타입을 int로 수정했습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao user 객체 메서드명 수정

리뷰 토대로 selectByUserEmail에서 selectUserByEmail로 메서드명을 수정했습니다.

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 이메일 & 닉네임 중복 체크 메서드명 변경

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 userDao의 이메일 & 닉네임 중복 체크 메서드명 변경

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 회원가입 메서드 수정

1. @DisplayName추가
2. 회원가입 테스트 메서드명 수정
3. selectUserByUserEmail에서 가져오는 객체를 담는 result변수를 findUser로 변경

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 이메일 중복 체크 메서드 수정

1. @DisplayName추가
2. 이메일 중복 체크 테스트 메서드명 수정
3. assertThat을 이용한 테스트 결과 형태 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 Dao 테스트 닉네임 중복 체크 메서드 수정

1. @DisplayName추가
2. 닉네임 중복 체크 테스트 메서드명 수정
3. assertThat을 이용한 테스트 결과 형태 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 회원가입 메서드 수정

1. @DisplayName추가
2. 회원가입 테스트 메서드명 수정
3. selectUserByUserEmail에서 가져오는 객체를 담는 result변수를 findUser로 변경

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 이메일 중복 체크 메서드 수정

1. @DisplayName추가
2. 이메일 중복 체크 테스트 메서드명 수정
3. assertThat을 이용한 테스트 결과 형태 수정

zlzzlzz2l committed:
Refactor: #10 - 회원 서비스 테스트 닉네임 중복 체크 메서드 수정

1. @DisplayName추가
2. 닉네임 중복 체크 테스트 메서드명 수정
3. assertThat을 이용한 테스트 결과 형태 수정

0xe82de reviewed (approved):
고생하셨습니다!!!

yjs1213 reviewed (approved):
고생하셨습니다~
